lib/spack/llnl/util/symlink.py

```
@@ -9,59 +9,136 @@
     else:
-        try:
-            # Try to use junctions
-            _win32_junction(real_path, link_path)
-        except OSError:
-            # If all else fails, fall back to copying files
-            shutil.copyfile(real_path, link_path)
+        # If windows can not make normal symbolic links
+        # we try junction for a directory or hardlink
+        # for a file.
+        if not os.path.isabs(link_path):
+            link_path = os.path.abspath(link_path)
+        # os.symlink will fail if link exists, emulate the behavior here
+        if exists(link_path):
+            raise OSError(errno.EEXIST, "Link exists: %s" % (link_path))
+        else:
+            windows_non_symlink(real_path, link_path)
+            if not os.path.exists(link_path):
+                raise OSError("Failed to create link: %s" % (link_path))
 
 
 def islink(path):
-    return os.path.islink(path) or _win32_is_junction(path)
+    """
+    Override os.islink to give correct answer for spack logic
+    """
+    if not is_windows:
+        return os.path.islink(path)
+
+    if windows_can_symlink():
+        return os.path.islink(path)
+
+    return windows_is_junction(path) or windows_is_hardlink(path)
+
 
+def windows_is_hardlink(path):
+    """
+    Determines if a path is a windows hardlink
+    """
+    if not is_windows or os.path.islink(path) or not os.path.exists(path):
+        return False
+
+    try:
+        cmd = ["fsutil", "hardlink", "list", path]
+        ret = subprocess.check_output(cmd)
+        lines = ret.decode().splitlines()
+        # We expect output of fsutil call to have at least
+        # two lines if the path is a hardlink
+        if len(lines) == 1:
+            return False
+        elif len(lines) > 1:
+            return True
+        else:
+            console.msg("[symlink] Cannot determine if hardlink. Returning false.")
+            return False
```

A couple points:

* Is there a way to do this by examining the file attributes as is done in `windows_is_junction` (i.e. with `GetFileAttributes`), rather than calling an external executable?
* If not, then I'm hoping we can conduct a more-robust check than examining the number of lines (e.g. does the output include the name of the file if it is a hard link). Moreover, it would be good to examine the exit code (if non-zero, we can raise an exception)
* Finally, do you know when we would end up in this `else` statement? If not, it's probably best to just raise an exception here (and not print a debug message to the terminal).

---

Now using os.stat(path).st_nlink to get the number of links and raising a new SymlinkError exception if the call fails